File: server/audio/whisper_sidecar.py
#!/usr/bin/env python3
"""Whisper sidecar — long-running process for audio transcription.

Reads JSON Lines from stdin, transcribes audio with Whisper, writes JSON Lines to stdout.

Protocol:
  Request:  {"id": "req-1", "audio_base64": "UklGR...", "language": "de", "model": "large-v3"}
  Progress: {"id": "req-1", "progress": true, "chunk": 1, "total": 3, "text": "partial text"}
  Response: {"id": "req-1", "text": "transkribierter text"}
  Error:    {"id": "req-1", "error": "reason"}

Features:
  - Chunked transcription for long audio (>60s split into 60s segments)
  - Progress updates per chunk
  - Model selection per request (large-v3, turbo, medium, distil-large-v3)
"""

import sys
import json
import base64
import tempfile
import os
import wave
import struct

# Chunk duration in seconds for long audio
CHUNK_DURATION_SECS = 60

def get_wav_duration(wav_path):
    """Get duration of a WAV file in seconds."""
    try:
        with wave.open(wav_path, 'rb') as w:
            frames = w.getnframes()
            rate = w.getframerate()
            if rate == 0:
                return 0
            return frames / rate
    except Exception:
        return 0

def split_wav(wav_path, chunk_secs=CHUNK_DURATION_SECS):
    """Split a WAV file into chunks of chunk_secs duration. Returns list of temp file paths."""
    try:
        with wave.open(wav_path, 'rb') as w:
            params = w.getparams()
            total_frames = w.getnframes()
            chunk_frames = int(params.framerate * chunk_secs)

            if total_frames <= chunk_frames * 1.2:
                # Audio is short enough — don't split (allow 20% over)
                return [wav_path]

            chunks = []
            offset = 0
            while offset < total_frames:
                end = min(offset + chunk_frames, total_frames)
                w.setpos(offset)
                data = w.readframes(end - offset)

                tmp_fd, tmp_path = tempfile.mkstemp(suffix=".wav")
                with wave.open(tmp_path, 'wb') as out:
                    out.setparams(params)
                    out.writeframes(data)
                os.close(tmp_fd)
                chunks.append(tmp_path)
                offset = end

            return chunks
    except Exception:
        return [wav_path]

def main():
    sys.stdout.reconfigure(line_buffering=True)

    sys.stderr.write("[whisper-sidecar] Starting up...\n")
    sys.stderr.flush()

    try:
        import whisper
        import torch
    except ImportError as e:
        sys.stderr.write(f"[whisper-sidecar] Missing dependency: {e}\n")
        sys.stderr.write("[whisper-sidecar] Install with: pip3 install openai-whisper torch\n")
        sys.stderr.flush()
        sys.exit(1)

    device = "mps" if torch.backends.mps.is_available() else "cpu"
    sys.stderr.write(f"[whisper-sidecar] Using device: {device}\n")
    sys.stderr.flush()

    # ── MLX fast path (Apple Silicon) ────────────────────────────────────────
    # mlx-whisper is ~3-4x faster than openai-whisper on M-series chips
    # (benchmark 2026-07-12, large-v3-turbo, 10.8s German clip: 0.29s vs
    # 1.06s warm on MPS). openai-whisper stays as the fallback for unmapped
    # model names, missing package, or MLX runtime errors.
    MLX_REPOS = {
        "large-v3-turbo": "mlx-community/whisper-large-v3-turbo",
        "turbo": "mlx-community/whisper-large-v3-turbo",
        "large-v3": "mlx-community/whisper-large-v3-mlx",
        "medium": "mlx-community/whisper-medium-mlx",
        "distil-large-v3": "mlx-community/distil-whisper-large-v3",
    }
    try:
        import mlx_whisper  # type: ignore
        mlx_available = True
        sys.stderr.write("[whisper-sidecar] mlx-whisper available — using MLX fast path.\n")
    except ImportError:
        mlx_whisper = None
        mlx_available = False
        sys.stderr.write("[whisper-sidecar] mlx-whisper not installed — using openai-whisper.\n")
    sys.stderr.flush()

    # Cache loaded torch models to avoid reloading on every request
    models = {}

    def get_model(model_name):
        if model_name not in models:
            sys.stderr.write(f"[whisper-sidecar] Loading model {model_name}...\n")
            sys.stderr.flush()
            models[model_name] = whisper.load_model(model_name, device=device)
            sys.stderr.write(f"[whisper-sidecar] Model {model_name} loaded.\n")
            sys.stderr.flush()
        return models[model_name]

    def transcribe_file(path, model_name, language):
        """MLX first (when available and the model is mapped), torch fallback."""
        if mlx_available and model_name in MLX_REPOS:
            try:
                result = mlx_whisper.transcribe(path, path_or_hf_repo=MLX_REPOS[model_name], language=language)
                return result.get("text", "").strip()
            except Exception as e:  # noqa: BLE001 — any MLX failure falls back to torch
                sys.stderr.write(f"[whisper-sidecar] MLX failed ({e}); falling back to openai-whisper.\n")
                sys.stderr.flush()
        result = get_model(model_name).transcribe(path, language=language, fp16=False)
        return result.get("text", "").strip()

    default_model = "large-v3-turbo"
    if mlx_available:
        # Prewarm MLX weights with a tiny silent clip so the first real
        # dictation doesn't pay the model-load cost.
        try:
            _fd, _warm = tempfile.mkstemp(suffix=".wav")
            with os.fdopen(_fd, "wb") as _f:
                with wave.open(_f, "wb") as w:
                    w.setnchannels(1)
                    w.setsampwidth(2)
                    w.setframerate(16000)
                    w.writeframes(b"\x00\x00" * 1600)  # 0.1s silence
            transcribe_file(_warm, default_model, "de")
            os.unlink(_warm)
            sys.stderr.write("[whisper-sidecar] MLX prewarmed.\n")
        except Exception as e:  # noqa: BLE001
            sys.stderr.write(f"[whisper-sidecar] Prewarm skipped: {e}\n")
        sys.stderr.flush()
    else:
        # Pre-load default torch model
        get_model(default_model)

    sys.stderr.write("[whisper-sidecar] Ready for requests.\n")
    sys.stderr.flush()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        try:
            req = json.loads(line)
        except json.JSONDecodeError:
            continue

        req_id = req.get("id", "unknown")
        audio_b64 = req.get("audio_base64", "")
        language = req.get("language", "de")
        model_name = req.get("model") or default_model

        if not audio_b64:
            print(json.dumps({"id": req_id, "error": "No audio data provided"}))
            continue

        # Decode audio to temp file
        tmp_fd, tmp_path = tempfile.mkstemp(suffix=".wav")
        try:
            audio_bytes = base64.b64decode(audio_b64)
            os.write(tmp_fd, audio_bytes)
            os.close(tmp_fd)

            duration = get_wav_duration(tmp_path)
            sys.stderr.write(f"[whisper-sidecar] {req_id}: {duration:.1f}s audio, model={model_name}\n")
            sys.stderr.flush()

            # Split into chunks if long audio
            chunks = split_wav(tmp_path)
            total_chunks = len(chunks)

            if total_chunks == 1:
                # Short audio — transcribe directly
                text = transcribe_file(chunks[0], model_name, language)
                print(json.dumps({"id": req_id, "text": text}))
            else:
                # Long audio — transcribe chunks with progress
                sys.stderr.write(f"[whisper-sidecar] {req_id}: split into {total_chunks} chunks\n")
                sys.stderr.flush()
                all_text = []
                for i, chunk_path in enumerate(chunks):
                    chunk_text = transcribe_file(chunk_path, model_name, language)
                    all_text.append(chunk_text)
                    # Send progress update
                    print(json.dumps({
                        "id": req_id,
                        "progress": True,
                        "chunk": i + 1,
                        "total": total_chunks,
                        "text": chunk_text,
                    }))
                    # Clean up chunk temp file (not the original)
                    if chunk_path != tmp_path:
                        try:
                            os.unlink(chunk_path)
                        except OSError:
                            pass

                # Send final merged result
                merged = " ".join(all_text)
                print(json.dumps({"id": req_id, "text": merged}))

        except Exception as e:
            print(json.dumps({"id": req_id, "error": str(e)}))
        finally:
            try:
                os.unlink(tmp_path)
            except OSError:
                pass


if __name__ == "__main__":
    main()

File: server/src/audio/whisper-sidecar.ts
import { spawn, ChildProcess } from 'child_process';
import path from 'path';
import { logger } from '../utils/logger';

interface PendingRequest {
  resolve: (text: string) => void;
  reject: (err: Error) => void;
  timer: NodeJS.Timeout;
  onProgress?: (info: { chunk: number; total: number; text: string }) => void;
}

// Watchdog: abort a request only if a single chunk makes no progress for this long.
// Each progress message resets the timer, so total audio length is irrelevant.
const CHUNK_STALL_TIMEOUT_MS = 45_000;
const SIDECAR_START_TIMEOUT_MS = 90_000;

// Find the server root (directory containing package.json) by walking up from __dirname.
function findServerRoot(): string {
  let dir = __dirname;
  for (let i = 0; i < 10; i++) {
    if (require('fs').existsSync(path.join(dir, 'package.json'))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return path.resolve(__dirname, '..', '..');
}

const SERVER_ROOT = findServerRoot();
const SIDECAR_DIR = path.join(SERVER_ROOT, 'audio');
const SIDECAR_SCRIPT = path.join(SIDECAR_DIR, 'whisper_sidecar_mlx.py');
const VENV_PYTHON = path.join(SIDECAR_DIR, '.venv-mlx', 'bin', 'python3');

let sidecar: ChildProcess | null = null;
let lineBuffer = '';
let requestId = 0;
const pending = new Map<string, PendingRequest>();
let startPromise: Promise<void> | null = null;

function ensureRunning(): Promise<void> {
  if (sidecar && !sidecar.killed) return Promise.resolve();
  if (startPromise) return startPromise;

  startPromise = new Promise<void>((resolve, reject) => {
    logger.info('[whisper] Starting sidecar...');

    const fs = require('fs');
    const pythonBin = fs.existsSync(VENV_PYTHON) ? VENV_PYTHON : 'python3';
    logger.info(`[whisper] Using Python: ${pythonBin}`);

    const child = spawn(pythonBin, [SIDECAR_SCRIPT], {
      stdio: ['pipe', 'pipe', 'pipe'],
    });

    let resolved = false;

    const startTimer = setTimeout(() => {
      if (resolved) return;
      resolved = true;
      logger.warn(`[whisper] Sidecar start timed out after ${SIDECAR_START_TIMEOUT_MS / 1000}s; killing`);
      try { child.kill('SIGKILL'); } catch {}
      sidecar = null;
      startPromise = null;
      reject(new Error(`Whisper sidecar start timed out (${SIDECAR_START_TIMEOUT_MS / 1000}s)`));
    }, SIDECAR_START_TIMEOUT_MS);

    child.stderr?.on('data', (chunk: Buffer) => {
      const text = chunk.toString();
      logger.info(`[whisper] ${text.trim()}`);
      if (!resolved && text.includes('Ready for requests')) {
        resolved = true;
        clearTimeout(startTimer);
        startPromise = null;
        sidecar = child;
        resolve();
      }
    });

    child.stdout?.on('data', (chunk: Buffer) => {
      lineBuffer += chunk.toString();
      const lines = lineBuffer.split('\n');
      lineBuffer = lines.pop() ?? '';
      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const resp = JSON.parse(line);
          const id = resp.id as string;
          const req = pending.get(id);
          if (!req) continue;

          // Progress update (chunk completed but more to come) — reset the watchdog.
          if (resp.progress) {
            clearTimeout(req.timer);
            req.timer = setTimeout(() => {
              pending.delete(id);
              req.reject(new Error(`Transkription Timeout (${CHUNK_STALL_TIMEOUT_MS / 1000}s). Chunk haengt.`));
            }, CHUNK_STALL_TIMEOUT_MS);
            req.onProgress?.({ chunk: resp.chunk, total: resp.total, text: resp.text ?? '' });
            continue; // Don't resolve yet — more chunks coming
          }

          pending.delete(id);
          clearTimeout(req.timer);
          if (resp.error) {
            req.reject(new Error(resp.error));
          } else {
            req.resolve(resp.text ?? '');
          }
        } catch {
          // ignore malformed lines
        }
      }
    });

    child.on('exit', (code) => {
      logger.warn(`[whisper] Sidecar exited with code ${code}`);
      clearTimeout(startTimer);
      sidecar = null;
      startPromise = null;
      for (const [, req] of pending) {
        clearTimeout(req.timer);
        req.reject(new Error('Whisper sidecar exited unexpectedly'));
      }
      pending.clear();
      if (!resolved) {
        resolved = true;
        reject(new Error('Whisper sidecar failed to start'));
      }
    });

    child.on('error', (err) => {
      logger.error(`[whisper] Failed to spawn sidecar: ${err.message}`);
      clearTimeout(startTimer);
      sidecar = null;
      startPromise = null;
      if (!resolved) {
        resolved = true;
        reject(new Error(`Whisper nicht verfuegbar: ${err.message}. Installieren mit: pip3 install openai-whisper torch`));
      }
    });
  });

  return startPromise;
}

export interface TranscribeOptions {
  language?: string;
  model?: string; // 'large-v3' | 'turbo' | 'medium'
  onProgress?: (info: { chunk: number; total: number; text: string }) => void;
}

export async function transcribe(audioBase64: string, options: TranscribeOptions = {}): Promise<string> {
  await ensureRunning();

  if (!sidecar?.stdin?.writable) {
    throw new Error('Whisper sidecar is not running');
  }

  const id = `req-${++requestId}`;

  logger.info(`[whisper] Transcription request ${id}: ${(audioBase64.length / 1024).toFixed(0)} KB Base64, stallTimeout=${CHUNK_STALL_TIMEOUT_MS / 1000}s, model=${options.model ?? 'default'}`);

  return new Promise<string>((resolve, reject) => {
    const timer = setTimeout(() => {
      pending.delete(id);
      reject(new Error(`Transkription Timeout (${CHUNK_STALL_TIMEOUT_MS / 1000}s). Chunk haengt oder Model zu langsam.`));
    }, CHUNK_STALL_TIMEOUT_MS);

    pending.set(id, { resolve, reject, timer, onProgress: options.onProgress });

    const request = JSON.stringify({
      id,
      audio_base64: audioBase64,
      language: options.language ?? 'de',
      model: options.model,
    }) + '\n';
    sidecar!.stdin!.write(request);
  });
}

export function shutdown(): void {
  if (sidecar && !sidecar.killed) {
    sidecar.kill('SIGTERM');
    sidecar = null;
  }
}
